src/main/java/com/google/devtools/build/lib/actions/cache/CompactPersistentActionCache.java

```diff
@@ -289,32 +316,162 @@ public void put(String key, ActionCache.Entry entry) {
+
+      // locationIndex
+      VarInt.putVarInt(value.getLocationIndex(), sink);
+
+      // actionId
+      byte[] actionIdBytes = value.getActionId().getBytes(ISO_8859_1);
+      VarInt.putVarInt(actionIdBytes.length, sink);
+      sink.write(actionIdBytes);
+
+      return sink.toByteArray();
+    } catch (IOException e) {
+      // This Exception can never be thrown by ByteArrayOutputStream.
+      throw new AssertionError(e);
+    }
+  }
+
+  private static RemoteFileArtifactValue decodeRemoteMetadata(byte[] data) throws IOException {
+    try {
+      ByteBuffer source = ByteBuffer.wrap(data);
+
+      // type
+      int type = VarInt.getVarInt(source);
+      Preconditions.checkState(type == MetadataType.Remote.id, "Invalid metadata type");
+
+      // digest
+      byte[] digest = new byte[VarInt.getVarInt(source)];
+      source.get(digest);
+
+      // size
+      long size = VarInt.getVarLong(source);
+
+      // locationIndex
+      int locationIndex = VarInt.getVarInt(source);
+
+      // actionId
+      byte[] actionIdBytes = new byte[VarInt.getVarInt(source)];
+      source.get(actionIdBytes);
+      String actionId = new String(actionIdBytes, ISO_8859_1);
+
+      if (source.remaining() > 0) {
+        throw new IOException("serialized metadata has not been fully decoded");
+      }
+
+      return new RemoteFileArtifactValue(digest, size, locationIndex, actionId);
+    } catch (BufferUnderflowException e) {
+      throw new IOException("encoded metadata is incomplete", e);
+    }
+  }
+
+  @Override
+  public void putFileMetadata(Artifact artifact, FileArtifactValue metadata) {
+    Preconditions.checkArgument(
+        !artifact.isTreeArtifact() && !artifact.isChildOfDeclaredDirectory(),
+        "Must use putTreeMetadata to save tree artifacts and their children: %s",
+        artifact
+    );
+    String key = artifact.getExecPathString();
+    byte[] content = encodeMetadata(metadata);
+    put(metadataMap, key, content);
+  }
+
+  @Override
+  public synchronized void removeFileMetadata(Artifact artifact) {
+    Preconditions.checkArgument(
+        !artifact.isTreeArtifact() && !artifact.isChildOfDeclaredDirectory(),
+        "Must use removeTreeMetadata to remote tree artifacts and their children: %s",
+        artifact
+    );
+    String key = artifact.getExecPathString();
+    metadataMap.remove(indexer.getIndex(key));
+  }
+
+  @Override
+  public FileArtifactValue getFileMetadata(Artifact artifact) {
+    Preconditions.checkArgument(
+        !artifact.isTreeArtifact() && !artifact.isChildOfDeclaredDirectory(),
+        "Must use getTreeMetadata to get tree artifacts and their children: %s",
+        artifact
+    );
+    String key = artifact.getExecPathString();
+    byte[] data = get(metadataMap, key);
+    try {
+      return data != null ? decodeMetadata(data) : null;
+    } catch (IOException e) {
+      // return null if can't decode metadata
+      return null;
```

Comment:
Maybe log some warning? I assume that would be an abnormal situation.

Reply:
Done.